ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java

```diff
@@ -347,97 +368,122 @@ private int bulkDeleteMetadata(List<CloudBlobMetadata> blobMetadataList) throws
+          // return null if exception is due to blob not being found
+          return null;
+        }
+        // For all other exceptions, pass the exception downstream.
+        throw throwable instanceof CompletionException ? (CompletionException) throwable : new CompletionException(ex);
       }
-      throw cex;
-    }
+      // return the blob metadata
+      return response.getItem();
+    }));
   }
 
   /**
-   * Update the blob metadata document in the CosmosDB collection.
+   * Update the blob metadata document in the CosmosDB collection asynchronously.
    * @param blobId the {@link BlobId} for which metadata is replaced.
    * @param updateFields Map of field names and new values to update.
-   * @return the {@link CosmosItemResponse} containing {@link CloudBlobMetadata} object returned by the operation, if successful.
-   * Returns Null if the field already has the specified value.
-   * @throws CosmosException if the record was not found or if the operation failed.
+   * @return a {@link CompletableFuture} that will eventually contain the {@link CloudBlobMetadata} returned by the
+   *         operation, if successful or {@code Null} if the field already has the specified value. If the record is not
+   *         found or an error occurred, it will contain a {@link CosmosException}.
    */
-  CosmosItemResponse<CloudBlobMetadata> updateMetadata(BlobId blobId, Map<String, String> updateFields)
-      throws CosmosException {
+  CompletableFuture<CloudBlobMetadata> updateMetadataAsync(BlobId blobId, Map<String, String> updateFields) {
+
+    CompletableFuture<CloudBlobMetadata> resultFuture = new CompletableFuture<>();
+    Timer.Context documentReadTimer = azureMetrics.documentReadTime.time();
+    AtomicReference<Timer.Context> documentUpdateTimer = new AtomicReference<>();
 
     // Read the existing record
-    CosmosItemResponse<CloudBlobMetadata> cosmosItemResponse = executeCosmosAction(
-        () -> cosmosAsyncContainer.readItem(blobId.getID(), new PartitionKey(blobId.getPartition().toPathString()),
-            CloudBlobMetadata.class).block(), azureMetrics.documentReadTime);
-    CloudBlobMetadata receivedMetadata = cosmosItemResponse.getItem();
-
-    // Update only if value has changed
-    Map<String, String> receivedFields = receivedMetadata.toMap();
-    Map<String, String> fieldsToUpdate = updateFields.entrySet()
-        .stream()
-        .filter(map -> !String.valueOf(updateFields.get(map.getKey())).equals(receivedFields.get(map.getKey())))
-        .collect(Collectors.toMap(Map.Entry::getKey, map -> String.valueOf(map.getValue())));
-    if (fieldsToUpdate.size() == 0) {
-      logger.debug("No change in value for {} in blob {}", updateFields.keySet(), blobId.getID());
-      return null;
-    }
+    cosmosAsyncContainer.readItem(blobId.getID(), new PartitionKey(blobId.getPartition().toPathString()),
+        CloudBlobMetadata.class).toFuture().exceptionally(throwable -> {
```

Comment:
if we just want to call documentReadTimer.stop when we finishes it, can we just use whenComplete instead of exceptionally, so 
1. we don't have to wrap and unwrap exception
2. we don't have to stop readtimer in thencompose function
3. we can complete resultFuture earlier.